products/servicedirectory/api.yaml

```diff
@@ -0,0 +1,179 @@
+    description: |
+      An individual service. A service contains a name and optional metadata.
+    parameters:
+      - !ruby/object:Api::Type::String
+        name: 'namespace'
+        description: |
+          The resource name of the namespace this service will belong to.
+        required: true
+        url_param_only: true
+      - !ruby/object:Api::Type::String
+        name: serviceId
+        description: |
+          The Resource ID must be 1-63 characters long, including digits,
+          lowercase letters or the hyphen character.
+        required: true
+        input: true
+        url_param_only: true
+    properties:
+      - !ruby/object:Api::Type::String
+        name: 'name'
+        description: |
+          The resource name for the service in the
+          format `projects/*/locations/*/namespaces/*/services/*`.
+        output: true
+      - !ruby/object:Api::Type::KeyValuePairs
+        name: 'metadata'
+        description: |
+          Metadata for the service. This data can be consumed
+          by service clients. The entire metadata dictionary may contain
+          up to 2000 characters, spread across all key-value pairs.
+          Metadata that goes beyond any these limits will be rejected.
+  - !ruby/object:Api::Resource
+    name: 'Endpoint'
+    base_url: '{{name}}'
+    create_url: '{{service}}/endpoints?endpointId={{endpoint_id}}'
+    self_link: '{{name}}'
+    update_verb: :PATCH
+    update_mask: true
+    references: !ruby/object:Api::Resource::ReferenceLinks
+      guides:
+        'Configuring an endpoint': 'https://cloud.google.com/service-directory/docs/configuring-service-directory#configuring_an_endpoint'
+      api: 'https://cloud.google.com/service-directory/docs/reference/rest/v1beta1/projects.locations.namespaces.services.endpoints'
+    min_version: beta
+    description: |
+      An individual endpoint that provides a service.
+    parameters:
+      - !ruby/object:Api::Type::String
+        name: 'service'
+        description: |
+          The resource name of the service that this endpoint provides.
+        required: true
+        url_param_only: true
+      - !ruby/object:Api::Type::String
+        name: endpointId
+        description: |
+          The Resource ID must be 1-63 characters long, including digits,
+          lowercase letters or the hyphen character.
+        required: true
+        input: true
+        url_param_only: true
+    properties:
+      - !ruby/object:Api::Type::String
+        name: 'name'
+        description: |
+          The resource name for the endpoint in the format
+          `projects/*/locations/*/namespaces/*/services/*/endpoints/*`.
+        output: true
+      - !ruby/object:Api::Type::String
+        name: 'address'
+        description: |
+          An IPv4 or IPv6 address.
+      - !ruby/object:Api::Type::Integer
+        name: 'port'
+        description: |
+          Service Directory will reject values outside of [0, 65535].
```

Comment:
nit: this description doesn't actually say what `port` represents, just how to specify it. Also, since it has specific values, you can add a validation into terraform.yaml (same with any other field that has constraints)

Reply:
Updated the descriptions and added validation functions.